src/development/cplusplus-codeguide.rst

```diff
@@ -0,0 +1,296 @@
+
+.. contents:: Table of Contents
+
+An Example C++ Project
+======================
+
+We have created a `skeleton C++ project
+<https://gitlab.com/steve-ord/cpp_template>`_. Which should provide a full introduction to the various
+recommendations and requirements for the development of C++. The philosophy behind the development of this
+template was to demonstrate one way to meet the project guidelines. There are probably as many ways to organise C++ projects
+as there are developers there are sure to be some contraversial design descisions made. 
+
+This projects demonstrates a recommended
+project layout. It also demonstrates how to implement the following recommended C++ project development features. 
+
+* Continous integration (CI) setup using Gitlab 
+* Cmake as a build tool.
+* GoogleTest framework and example unit testing.
+* C++ linting using clang for stylistic errors.
+* Also test running under valgrind for memory errors.
+* gcov to measure test coverage 
+
+All building and testing is done within a docker container.
+
+Project Layout
+--------------
+Top Level
+^^^^^^^^^
+We have a top level of the project containing:
+
+CMakeLists.txt
+""""""""""""""
+        We have built this template using CMake. The structure of this file
+        will be discussed in `Building the Project`_.   
+
+LICENSE
+"""""""
+        All projects *must* have a license. We have include the recommended BSD
+        3 clause license template *please fill it in*.
+
+README.md
+"""""""""
+        Should at the very least provide a brief discription, installation
+        instructions, pre-requisites
+
+src
+"""
+        This is the top of `the source tree`. It does not have to be called src.
+        But we recommend that the source tree has its head here. The first
+        namespace being the level below this one.
+
+version.txt
+"""""""""""
+        In this template the version file is parsed by the CMakeLists file during the 
+        build process. This specific functionality is not required, but clear versioning is.
+        We recommend - but do not mandate `Semantic Versioning <https://semver.org>`_.
+
+
+The Source Tree
+^^^^^^^^^^^^^^^
+Our recomended source tree structure follows:
+
+.. literalinclude:: source_tree.txt
+        
+A number of design decisions have been made here:
+
+What To Do With Namespaces
+""""""""""""""""""""""""""
+The structure follows the namespaces. We have defined a *top* and a *nested*
+namespace and the directory structure follows this structure. This allows code
+to be grouped together easily by namespace. Also the installation assumes this
+structure also. Therefore headers are included using their full namespaces.
+This avoids pollution of the install tree.
+
+.. warning:: we have not decided yet whether all projects should have a top-level SKA namespace
+
+Headers
+"""""""
+Header files are included with source. We have done this to make it easier
```

Comment:
Thats not an easy sentence to parse. Can we rewrite to make it clearer? "..are in the same directory as the implementation (.cpp) files"